primitives/account/src/account/staking_contract/staker.rs

```diff
@@ -283,41 +285,65 @@ impl StakingContract {
         staker_address: &Address,
         value: Coin,
         tx_logger: &mut TransactionLog,
-    ) -> Result<(), AccountError> {
+    ) -> Result<AddStakeReceipt, AccountError> {
         // Get the staker.
         let mut staker = store.expect_staker(staker_address)?;
 
-        // Fail if the minimum stake would be violated for the non-retired funds (invariant 1).
-        Staker::enforce_min_stake(
-            staker.active_balance + value,
-            staker.inactive_balance,
-            staker.retired_balance,
-        )?;
+        // Add stake txs never violate minimum stake for the non-retired funds (invariant 1),
+        // because the intrinsic tx checks that value is >= min stake.
+        assert!(
+            Staker::enforce_min_stake(
+                staker.active_balance + value,
+                staker.inactive_balance,
+                staker.retired_balance,
+            )
+            .is_ok(),
+            "Add stake should never violate the min stake invariants"
+        );
 
         // All checks passed, not allowed to fail from here on!
 
-        // If we are delegating to a validator, we need to update it.
-        if let Some(validator_address) = &staker.delegation {
-            // Check that the delegation is still valid, i.e. the validator hasn't been deleted.
-            store.expect_validator(validator_address)?;
-            self.increase_stake_to_validator(store, validator_address, value);
-        }
-
         // Update the staker's and staking contract's balances.
-        staker.active_balance += value;
+        // We want to preferentially credit the active balance. Only if there is no active balance,
+        // then we will choose the balance with the most funds between inactive and retired.
+        let credited_balance = if !staker.active_balance.is_zero() {
+            staker.active_balance += value;
+            BalanceType::Active
+        } else if staker.inactive_balance >= staker.retired_balance {
+            staker.inactive_balance += value;
+            BalanceType::Inactive
+        } else {
+            staker.retired_balance += value;
+            BalanceType::Retired
+        };
         self.balance += value;
 
+        // Create the receipt.
+        let receipt = AddStakeReceipt {
+            credited_balance: credited_balance.clone(),
+        };
+
+        if credited_balance == BalanceType::Active {
+            // If we are delegating to a validator, we need to update it.
+            if let Some(validator_address) = &staker.delegation {
+                // Check that the delegation is still valid, i.e. the validator hasn't been deleted.
+                store.expect_validator(validator_address)?;
```

Comment:
There might be a problem with the ordering. We usually want to do changes only after all potentially failing conditions are checked.
In practice, I don't think it is a problem, as all previous changes are in the staker and that one needs to be put explicitly back.